lib/pubnub/client/paged_history.rb

```diff
@@ -1,60 +1,187 @@
+# frozen_string_literal: true
+
 # Toplevel Pubnub module.
 module Pubnub
   # Pubnub client Class
   class Client
     # Module that holds paged_history event logic
     module PagedHistory
+      # Fetch messages as long as one of criteria won't be met.
+      # Messages returned as single envelope.
+      def all_history_messages(options = {}, &block)
+        channel = options.fetch(:channel)
+        # Time frame between which messages should be fetched.
+        start_tt = options.fetch(:start, nil)
+        end_tt = options.fetch(:end, nil)
+        page_size = options.fetch(:page_size, 100)          # How many messages per-request should be fetched.
+        reverse = options.fetch(:reverse, false)            # Order in which messages should be retrieved if :end not set.
+        include_token = options.fetch(:include_token, true) # Whether timetoken should be included with message or not.
+        maximum = options.fetch(:max, 500)                  # Maximum number of messages which should be fetched.
+        callback = options.fetch(:callback, block)
+
+        reverse = false unless end_tt.nil? # Disable revers if closed time interval specified.
+        maximum = nil unless end_tt.nil? # Disable maximum messages count if closed time interval specified.
+
+        if options[:http_sync]
+          sync_all_history_messages(channel, include_token, page_size, reverse, maximum, callback, start: start_tt, end: end_tt)
+        else
+          async_all_history_messages(options)
```

Comment:
@kleewho the only purpose of `async` method version is to create `Feature` with execute closure (it will execute later, for example, when access `.value` of feature) and switch `http_sync` to `false` - in this case `options` can be again passed to `all_history_messages` which will call `sync_all_history_messages`

Reply:
Oh right. Now I get it